pymatviz/_data.py

```diff
@@ -0,0 +1,303 @@
+"""Utils for data preprocessing and related tasks."""
+
+from __future__ import annotations
+
+import warnings
+from collections.abc import Iterable, Sequence
+from typing import TYPE_CHECKING, Union, get_args
+
+import numpy as np
+import pandas as pd
+from pymatgen.core import Element
+
+from pymatviz.enums import Key
+
+
+if TYPE_CHECKING:
+    from typing import Literal
+
+
+# Data types used internally by ptable plotters
+SupportedValueType = Union[Sequence[float], np.ndarray]
+
+# Data types that can be passed to PTableProjector and data_preprocessor
+SupportedDataType = Union[
+    dict[str, Union[float, Sequence[float], np.ndarray]], pd.DataFrame, pd.Series
+]
+
+
+def check_for_missing_inf(df_in: pd.DataFrame, col: str) -> tuple[bool, bool]:
+    """Check if there is NaN or infinity in a DataFrame column.
+
+    Args:
+        df_in (pd.DataFrame): DataFrame to check.
+        col (str): Name of the column to check.
+
+    Returns:
+        tuple[bool, bool]: Has NaN, has infinity.
+    """
+    # Check if there is missing value or infinity
+    all_values = df_in[col].explode().explode().explode()
+
+    # Convert to numeric, forcing non-numeric types to NaN
+    all_values = pd.to_numeric(all_values, errors="coerce")
+
+    # Check for NaN
+    has_nan = False
+    if all_values.isna().to_numpy().any():
+        warnings.warn("NaN found in data", stacklevel=2)
+        has_nan = True
+
+    # Check for infinity
+    has_inf = False
+    if np.isinf(all_values).to_numpy().any():
+        warnings.warn("Infinity found in data", stacklevel=2)
+        has_inf = True
+
+    return has_nan, has_inf
+
+
+def get_df_nest_level(df_in: pd.DataFrame, col: str, max_level: int = 10) -> int:
+    """Check for maximum nest level in a DataFrame column.
+
+    Definition of nest level:
+        Level 0 (no list):       "Fe": 1
+        Level 1 (flat list):     "Co": [1, 2]
+        Level 2 (nested lists):  "Ni": [[1, 2], [3, 4], ]
+        ...
+
+    Args:
+        df_in (pd.DataFrame): The DataFrame to check.
+        col (str): Name of the column to check.
+        max_level (int): Max nest level to try up to. Defaults to 10.
+
+    Returns:
+        int: The maximum nest level.
+    """
+    current_level = 0
+    while current_level <= max_level:
+        # Check if any element in the column is a list
+        if df_in[col].map(lambda x: isinstance(x, list)).any():
```

Comment:
> what if the column values are arrays instead of lists? 

Great! I already fixed this and updated the unit test.

> also, maybe instead of looping it would be better to call `np.array()` on one of the lists and then return `arr.ndim`

This sounds like a wonderful solution. But I'm still trying to get it to work, because we might be expecting different dimensions for different elements

```python3
df_level_2 = pd.DataFrame(
    {
        "Fe": [1, 2, 3],
        "O": [[4, 5], [6, 7]],  # get max level
    }.items(),
    columns=[Key.element, Key.heat_val],
).set_index(Key.element)

column_values = df_level_2[Key.heat_val].tolist()
array = np.array(column_values)
```

Which raises:
```python3
Traceback (most recent call last):
  File "/Users/yang/developer/pymatviz/test_df.py", line 17, in <module>
    array = np.array(column_values)
            ^^^^^^^^^^^^^^^^^^^^^^^
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
```